### PR TITLE
rgw: replace null_yields with optional ones

### PR DESCRIPTION
**Pull request by TRYTOBE8TME**
## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/rgw/driver/rados/rgw_gc.cc [hunk 1]
```diff
@@ -51,7 +51,7 @@ void RGWGC::initialize(CephContext *_cct, RGWRados *_store) {
     op.create(false);
     const uint64_t queue_size = cct->_conf->rgw_gc_max_queue_size, num_deferred_entries = cct->_conf->rgw_gc_max_deferred;
     gc_log_init2(op, queue_size, num_deferred_entries);
-    store->gc_operate(this, obj_names[i], &op);
+    store->gc_operate(this, obj_names[i], &op, null_yield);
```
Comment (cbodley) [inline src/rgw/driver/rados/rgw_gc.cc:54]: you added an `optional_yield y` argument to `RGWGC::initialize()`, so you can pass that here. same with `RGWGC::send_split_chain()` and several other functions in rgw_gc.cc

**Review by cbodley — COMMENTED** [review]
looks great overall! 'make check' is catching some unit tests under src/test/rgw/ that need updating also

**Comment by dang** [comment]
BTW, you can "make tests" (or "ninja tests") to build the tests that make check uses without running them.

File: src/rgw/driver/rados/rgw_object_expirer_core.cc [hunk 2]
```diff
@@ -188,7 +188,7 @@ int RGWObjExpStore::objexp_hint_trim(const DoutPrefixProvider *dpp,
   }
   auto& ref = obj.get_ref();
   int ret = cls_timeindex_trim_repeat(dpp, ref, oid, utime_t(start_time), utime_t(end_time),
-          from_marker, to_marker);
+          from_marker, to_marker, null_yield);
```
Comment (alimaredia) [inline src/rgw/driver/rados/rgw_object_expirer_core.cc:191]: You added an optional_yield to `objexp_hint_trim` so you don't need a `null_yield` here.

File: src/rgw/driver/rados/rgw_object_expirer_core.cc [hunk 3]
```diff
@@ -273,7 +273,7 @@ void RGWObjectExpirer::trim_chunk(const DoutPrefixProvider *dpp,
   real_time rt_to = to.to_real_time();
 
   int ret = exp_store.objexp_hint_trim(dpp, shard, rt_from, rt_to,
-                                       from_marker, to_marker);
+                                       from_marker, to_marker, null_yield);
```
Comment (alimaredia) [inline src/rgw/driver/rados/rgw_object_expirer_core.cc:276]: Same here, there's an optional_yield added to `trim_chunk`.

File: src/rgw/driver/rados/rgw_object_expirer_core.cc [hunk 4]
```diff
@@ -327,7 +327,7 @@ bool RGWObjectExpirer::process_single_shard(const DoutPrefixProvider *dpp,
     garbage_chunk(dpp, entries, need_trim);
 
     if (need_trim) {
-      trim_chunk(dpp, shard, last_run, round_start, marker, out_marker);
+      trim_chunk(dpp, shard, last_run, round_start, marker, out_marker, null_yield);
```
Comment (alimaredia) [inline src/rgw/driver/rados/rgw_object_expirer_core.cc:330]: Same here, there's an optional_yield added to `process_single_shard`.

File: src/rgw/driver/rados/rgw_object_expirer_core.cc [hunk 5]
```diff
@@ -358,7 +358,7 @@ bool RGWObjectExpirer::inspect_all_shards(const DoutPrefixProvider *dpp,
 
     ldpp_dout(dpp, 20) << "processing shard = " << shard << dendl;
 
-    if (! process_single_shard(dpp, shard, last_run, round_start)) {
+    if (! process_single_shard(dpp, shard, last_run, round_start, null_yield)) {
```
Comment (alimaredia) [inline src/rgw/driver/rados/rgw_object_expirer_core.cc:361]: Same here, there's an optional_yield added to `inspect_all_shards`.

File: src/rgw/driver/rados/rgw_rados.cc [hunk 6]
```diff
@@ -7953,15 +7953,15 @@ int RGWRados::get_bucket_info(RGWServices *svc,
 
 int RGWRados::try_refresh_bucket_info(RGWBucketInfo& info,
                                       ceph::real_time *pmtime,
-                                      const DoutPrefixProvider *dpp,
+                                      const DoutPrefixProvider *dpp, optional_yield y,
```
Comment (alimaredia) [inline src/rgw/driver/rados/rgw_rados.cc:7956]: @cbodley It's a knit, but there are dozens of instances in this PR where the `optional_yield` argument in a function definition is added on the same line as another argument even though every other argument in the function is on it's own line. I think these optional_yields should go on their own line. What do you think?

File: src/rgw/driver/rados/rgw_rados.cc [hunk 7]
```diff
@@ -4586,7 +4586,7 @@ int RGWRados::copy_obj_data(RGWObjectCtx& obj_ctx,
   AtomicObjectProcessor processor(&aio, this->driver, &dest_placement,
                                   bucket->get_info().owner, obj_ctx,
                                   dest_obj->clone(), olh_epoch, tag,
-				  dpp, null_yield);
+				  dpp, y);
```
Comment (alimaredia) [inline src/rgw/driver/rados/rgw_rados.cc:4589]: @kapandya @cbodley just a heads up I added the comment on the lines above this AtomicObjectProcessor initialization saying that changing this `null_yield` to a `y` causes ragweed tests to fail. I wouldn't be surprised if it happens when we run this PR through teuthology again.

File: src/rgw/rgw_crypt.cc [hunk 8]
```diff
@@ -975,7 +975,7 @@ static int get_sse_s3_bucket_key(req_state *s,
       if (res != -ECANCELED) {
         break;
       }
-      res = s->bucket->try_refresh_info(s, nullptr);
+      res = s->bucket->try_refresh_info(s, nullptr, null_yield);
```
Comment (alimaredia) [inline src/rgw/rgw_crypt.cc:978]: you can use `s->yield` here and anywhere else there is a `req_state` which is commonly called `s`.

File: src/rgw/rgw_op.cc [hunk 9]
```diff
@@ -1261,14 +1261,14 @@ void RGWPutBucketReplication::execute(optional_yield y) {
 
     s->bucket->get_info().set_sync_policy(std::move(sync_policy));
 
-    int ret = s->bucket->put_info(this, false, real_time());
+    int ret = s->bucket->put_info(this, false, real_time(), null_yield);
```
Comment (alimaredia) [inline src/rgw/rgw_op.cc:1264]: This function takes a `y`.

File: src/rgw/rgw_op.cc [hunk 10]
```diff
@@ -1305,14 +1305,14 @@ void RGWDeleteBucketReplication::execute(optional_yield y)
...
```
Comment (alimaredia) [inline src/rgw/rgw_op.cc:1308]: This function takes a `y`.

File: src/rgw/rgw_op.cc [hunk 11]
```diff
@@ -2819,9 +2819,9 @@ void RGWSetBucketWebsite::execute(optional_yield y)
   op_ret = retry_raced_bucket_write(this, s->bucket.get(), [this] {
       s->bucket->get_info().has_website = true;
       s->bucket->get_info().website_conf = website_conf;
-      op_ret = s->bucket->put_info(this, false, real_time());
+      op_ret = s->bucket->put_info(this, false, real_time(), null_yield);
```
Comment (alimaredia) [inline src/rgw/rgw_op.cc:2822]: This function takes a `y`.

File: src/rgw/rgw_op.cc [hunk 12]
```diff
@@ -2862,9 +2862,9 @@ void RGWDeleteBucketWebsite::execute(optional_yield y)
   op_ret = retry_raced_bucket_write(this, s->bucket.get(), [this] {
       s->bucket->get_info().has_website = false;
       s->bucket->get_info().website_conf = RGWBucketWebsiteConf();
-      op_ret = s->bucket->put_info(this, false, real_time());
+      op_ret = s->bucket->put_info(this, false, real_time(), null_yield);
```
Comment (alimaredia) [inline src/rgw/rgw_op.cc:2865]: This function takes a `y`.

File: src/rgw/rgw_op.cc [hunk 13]
```diff
@@ -8305,9 +8305,9 @@ void RGWPutBucketObjectLock::execute(optional_yield y)
 
   op_ret = retry_raced_bucket_write(this, s->bucket.get(), [this] {
     s->bucket->get_info().obj_lock = obj_lock;
-    op_ret = s->bucket->put_info(this, false, real_time());
+    op_ret = s->bucket->put_info(this, false, real_time(), null_yield);
```
Comment (alimaredia) [inline src/rgw/rgw_op.cc:8308]: This function takes a `y`.

**Comment by alimaredia** [comment]
@kapandya great start to this work! On top of the other comments that were left do you have a list of which ops don't have anymore librados blocking warnings anymore and which still do or are untested? What happens with a full run of s3-tests?

**Comment by TRYTOBE8TME** [comment]
jenkins test make check

**Comment by TRYTOBE8TME** [comment]
Teuthology run: https://pulpito.ceph.com/kapandya-2023-03-29_14:12:29-rgw-wip-rgw-yield-work-distro-default-smithi/

File: src/rgw/driver/rados/rgw_putobj_processor.cc [hunk 14]
```diff
@@ -168,7 +168,7 @@ RadosWriter::~RadosWriter()
       continue;
     }
 
-    int r = store->delete_raw_obj(dpp, obj);
+    int r = store->delete_raw_obj(dpp, obj, null_yield);
```
Comment (dang) [inline src/rgw/driver/rados/rgw_putobj_processor.cc:171]: RadosWriter has an optional_yield in it, so this can be y.

**Review by dang — COMMENTED** [review]
Just one spot I see that needs updating

**Review by dang — APPROVED** [review]

File: src/rgw/driver/rados/rgw_notify.cc [hunk 15]
```diff
@@ -780,7 +780,7 @@ static inline bool notification_match(reservation_t& res,
   const RGWPubSub ps(res.store, res.user_tenant);
   const RGWPubSub::Bucket ps_bucket(ps, res.bucket);
   rgw_pubsub_bucket_topics bucket_topics;
-  auto rc = ps_bucket.get_topics(res.dpp, bucket_topics, res.yield);
+  auto rc = ps_bucket.get_topics(res.dpp, bucket_topics, res.s->yield);
```
Comment (cbodley) [inline src/rgw/driver/rados/rgw_notify.cc:783]: i think this is the cause of the crash you're seeing in teuthology. `RGWDeleteMultiObj::handle_individual_object()` runs under a different yield context than the req_state's `res.s->yield`

**Comment by TRYTOBE8TME** [comment]
Teuthology run: https://pulpito.ceph.com/kapandya-2023-04-28_06:37:55-rgw-wip-rgw-yield-work-distro-default-smithi/

**Comment by alimaredia** [comment]
I noticed this job in the verify suite failed: http://qa-proxy.ceph.com/teuthology/kapandya-2023-04-28_06:37:55-rgw-wip-rgw-yield-work-distro-default-smithi/7256679/teuthology.log

with:
2023-04-30T09:35:11.835 INFO:teuthology.orchestra.run.smithi161.stdout:s3tests_boto3/functional/test_s3.py::test_bucket_list_return_data_versioning FAILED [ 35%]

Is that a known failure? As of this run by @cbodley on `main` yesterday the verify suite was green: https://pulpito.ceph.com/cbodley-2023-04-26_18:44:09-rgw:verify-main-distro-default-smithi/

**Comment by mattbenjamin** [comment]
what's the current status after the latest rebase?  is it ready for merge?

**Comment by alimaredia** [comment]
@mattbenjamin on @kapandya's latest run there were s3-test failures in the rgw verify suite.

This [run](http://qa-proxy.ceph.com/teuthology/kapandya-2023-04-28_06:37:55-rgw-wip-rgw-yield-work-distro-default-smithi/7256714/teuthology.log) failed on: test_bucket_list_return_data

Thus [run](http://qa-proxy.ceph.com/teuthology/kapandya-2023-04-28_06:37:55-rgw-wip-rgw-yield-work-distro-default-smithi/7256679/teuthology.log) failed on: test_bucket_list_return_data_versioning

There are also s3test failures in the lifecycle suite (http://qa-proxy.ceph.com/teuthology/kapandya-2023-04-28_06:37:55-rgw-wip-rgw-yield-work-distro-default-smithi/7256663/teuthology.log) and in cloud transition suite (http://qa-proxy.ceph.com/teuthology/kapandya-2023-04-28_06:37:55-rgw-wip-rgw-yield-work-distro-default-smithi/7256659/teuthology.log).

@soumyakoduri are there known s3-test failures in the cloud transition suite?

**Comment by mattbenjamin** [comment]
@alimaredia is someone helping to verify and resolve these?

**Comment by TRYTOBE8TME** [comment]
This is my latest run: https://pulpito.ceph.com/kapandya-2023-05-09_05:23:55-rgw-wip-rgw-yield-work-distro-default-smithi/

**Comment by mattbenjamin** [comment]
What is your analysis of the failures in this run @TRYTOBE8TME ?

**Comment by soumyakoduri** [comment]
> This is my latest run: https://pulpito.ceph.com/kapandya-2023-05-09_05:23:55-rgw-wip-rgw-yield-work-distro-default-smithi/

The lifecycle and cloud-transition tests seem to be consistently failing. I would suggest to test lifecycle transition functionality locally first and verify if there is any regression.

**Comment by TRYTOBE8TME** [comment]
> > This [run](http://qa-proxy.ceph.com/teuthology/kapandya-2023-04-28_06:37:55-rgw-wip-rgw-yield-work-distro-default-smithi/7256714/teuthology.log) failed on: test_bucket_list_return_data
> > Thus [run](http://qa-proxy.ceph.com/teuthology/kapandya-2023-04-28_06:37:55-rgw-wip-rgw-yield-work-distro-default-smithi/7256679/teuthology.log) failed on: test_bucket_list_return_data_versioning
> 
> @TRYTOBE8TME are you able to reproduce these failures by running s3-tests locally? or do they only happen in teuthology?

@cbodley I checked it but it happens only in teuthology and not locally.

**Comment by mattbenjamin** [comment]
You need to learn to run s3 tests locally.  You can run just those tests with a correct invocation.  Try doing this against a vstart cluster.

**Comment by cbodley** [comment]
i also saw `test_bucket_list_return_data` fail once out of 20 runs, so it's not specific to versioned buckets. `_do_write_meta()` sets the same mtime on the head object that it writes to the bucket index. i'll try to debug further

**Comment by cbodley** [comment]
i added some extra logging for these timestamps, and caught the output from a test failure:
```
2023-05-19T13:13:13.994-0400 54f8c6c0 10 req 9230964859717032992 0.016000582s s3:put_obj _do_write_meta() generated object mtime 2023-05-19T13:13:13.995343-0400
2023-05-19T13:13:13.994-0400 54f8c6c0 10 req 9230964859717032992 0.016000582s s3:put_obj _do_write_meta() head object mtime sec=1684516393 nsec=995343547
...
2023-05-19T13:13:14.001-0400 4576d6c0  1 -- 192.168.245.128:0/1393891603 --> [v2:192.168.245.128:6800/2322402219,v1:192.168.245.128:6801/2322402219] -- osd_op(unknown.0.0:1729 6.0 6:50ccb7c4:::ec68f463-21c8-4252-8a5c-61f14aff422c.4191.29_foo:head [create,setxattr user.rgw.idtag (62) in=76b,setxattr user.rgw.tail_tag (62) in=79b,writefull 0~3 in=3b,setxattr user.rgw.manifest (364) in=381b,setxattr user.rgw.acl (147) in=159b,setxattr user.rgw.etag (32) in=45b,setxattr user.rgw.x-amz-content-sha256 (65) in=94b,setxattr user.rgw.x-amz-date (17) in=36b,call rgw.obj_store_pg_ver in=44b,setxattr user.rgw.source_zone (4) in=24b] snapc 0=[] ondisk+write+known_if_redirected+supports_pool_eio e19) v8 -- 0x3eba3f60 con 0x17ba8ea0
2023-05-19T13:13:14.004-0400 129806c0  1 -- 192.168.245.128:0/1393891603 <== osd.0 v2:192.168.245.128:6800/2322402219 2069 ==== osd_op_reply(1729 ec68f463-21c8-4252-8a5c-61f14aff422c.4191.29_foo [create,setxattr (62),setxattr (62),writefull 0~3,setxattr (364),setxattr (147),setxattr (32),setxattr (65),setxattr (17),call,setxattr (4)] v19'1161 uv1161 ondisk = 0) v8 ==== 612+0+0 (crc 0 0 0) 0x100cf190 con 0x17ba8ea0
...
2023-05-19T13:13:14.034-0400 4676f6c0 15 req 7446894691545786343 0.012000435s Encryption mode:
2023-05-19T13:13:14.034-0400 4676f6c0 10 req 7446894691545786343 0.012000435s s3:get_obj dumping Last-Modified 2023-05-19T13:13:14.002009-0400
2023-05-19T13:13:14.035-0400 4676f6c0  2 req 7446894691545786343 0.013000472s s3:get_obj completing
...
2023-05-19T13:13:14.109-0400 477716c0  2 req 3206062128404960865 0.013000472s s3:list_bucket completing
2023-05-19T13:13:14.110-0400 477716c0 10 req 3206062128404960865 0.014000508s s3:list_bucket dumping LastModified 2023-05-19T13:13:13.995343-0400
```
`_do_write_meta()` calls `op.mtime2()` to set the head object's mtime, but that doesn't appear to take effect; when get_obj reads the head object back, its gets a more recent mtime. presumably because the osd set it to the current time instead

**Comment by TRYTOBE8TME** [comment]
> i added some extra logging for these timestamps, and caught the output from a test failure:
> 
> ```
> 2023-05-19T13:13:13.994-0400 54f8c6c0 10 req 9230964859717032992 0.016000582s s3:put_obj _do_write_meta() generated object mtime 2023-05-19T13:13:13.995343-0400
> 2023-05-19T13:13:13.994-0400 54f8c6c0 10 req 9230964859717032992 0.016000582s s3:put_obj _do_write_meta() head object mtime sec=1684516393 nsec=995343547
> ...
> 2023-05-19T13:13:14.001-0400 4576d6c0  1 -- 192.168.245.128:0/1393891603 --> [v2:192.168.245.128:6800/2322402219,v1:192.168.245.128:6801/2322402219] -- osd_op(unknown.0.0:1729 6.0 6:50ccb7c4:::ec68f463-21c8-4252-8a5c-61f14aff422c.4191.29_foo:head [create,setxattr user.rgw.idtag (62) in=76b,setxattr user.rgw.tail_tag (62) in=79b,writefull 0~3 in=3b,setxattr user.rgw.manifest (364) in=381b,setxattr user.rgw.acl (147) in=159b,setxattr user.rgw.etag (32) in=45b,setxattr user.rgw.x-amz-content-sha256 (65) in=94b,setxattr user.rgw.x-amz-date (17) in=36b,call rgw.obj_store_pg_ver in=44b,setxattr user.rgw.source_zone (4) in=24b] snapc 0=[] ondisk+write+known_if_redirected+supports_pool_eio e19) v8 -- 0x3eba3f60 con 0x17ba8ea0
> 2023-05-19T13:13:14.004-0400 129806c0  1 -- 192.168.245.128:0/1393891603 <== osd.0 v2:192.168.245.128:6800/2322402219 2069 ==== osd_op_reply(1729 ec68f463-21c8-4252-8a5c-61f14aff422c.4191.29_foo [create,setxattr (62),setxattr (62),writefull 0~3,setxattr (364),setxattr (147),setxattr (32),setxattr (65),setxattr (17),call,setxattr (4)] v19'1161 uv1161 ondisk = 0) v8 ==== 612+0+0 (crc 0 0 0) 0x100cf190 con 0x17ba8ea0
> ...
> 2023-05-19T13:13:14.034-0400 4676f6c0 15 req 7446894691545786343 0.012000435s Encryption mode:
> 2023-05-19T13:13:14.034-0400 4676f6c0 10 req 7446894691545786343 0.012000435s s3:get_obj dumping Last-Modified 2023-05-19T13:13:14.002009-0400
> 2023-05-19T13:13:14.035-0400 4676f6c0  2 req 7446894691545786343 0.013000472s s3:get_obj completing
> ...
> 2023-05-19T13:13:14.109-0400 477716c0  2 req 3206062128404960865 0.013000472s s3:list_bucket completing
> 2023-05-19T13:13:14.110-0400 477716c0 10 req 3206062128404960865 0.014000508s s3:list_bucket dumping LastModified 2023-05-19T13:13:13.995343-0400
> ```
> 
> `_do_write_meta()` calls `op.mtime2()` to set the head object's mtime, but that doesn't appear to take effect; when get_obj reads the head object back, its gets a more recent mtime. presumably because the osd set it to the current time instead

@cbodley can you please give me some idea on how this can be fixed?

**Comment by cbodley** [comment]
@TRYTOBE8TME i tracked the bug down to the librados client. our call to `op.mtime2()` was only being applied by synchronous calls to `IoCtx::operate()`, and not to the async calls to `IoCtx::aio_operate()` now that a yield context is present

i opened https://tracker.ceph.com/issues/61349 and https://github.com/ceph/ceph/pull/51681 with the fix. i'd like to get that reviewed and tested separately

in the meantime, can you please rebase this PR and add that commit on top for testing?

**Comment by mattbenjamin** [comment]
@cbodley @adamemerson thanks sincerely for chasing this down

**Comment by TRYTOBE8TME** [comment]
Latest run: https://pulpito.ceph.com/kapandya-2023-05-23_10:05:30-rgw-wip-rgw-yield-check-final-distro-default-smithi/

**Comment by alimaredia** [comment]
> Latest run: https://pulpito.ceph.com/kapandya-2023-05-23_10:05:30-rgw-wip-rgw-yield-check-final-distro-default-smithi/

That run looks cleaner but there are still a few failures I'm not sure are in the baseline:
This [run](http://qa-proxy.ceph.com/teuthology/kapandya-2023-05-23_10:05:30-rgw-wip-rgw-yield-check-final-distro-default-smithi/7283525/teuthology.log) fails running rgw-reshard.sh in the workunits

The [run](http://qa-proxy.ceph.com/teuthology/kapandya-2023-05-23_10:05:30-rgw-wip-rgw-yield-check-final-distro-default-smithi/7283499/teuthology.log) fails an s3-test in the crypt suite.

There are also some valgrind failures. Do any of these seem familiar @cbodley

**Comment by cbodley** [comment]
> > Latest run: https://pulpito.ceph.com/kapandya-2023-05-23_10:05:30-rgw-wip-rgw-yield-check-final-distro-default-smithi/
> 
> That run looks cleaner but there are still a few failures I'm not sure are in the baseline: This [run](http://qa-proxy.ceph.com/teuthology/kapandya-2023-05-23_10:05:30-rgw-wip-rgw-yield-check-final-distro-default-smithi/7283525/teuthology.log) fails running rgw-reshard.sh in the workunits

that looks like a test case is failing: https://github.com/ceph/ceph/blob/62d1cc0/qa/workunits/rgw/test_rgw_reshard.py#L209-L212

the `IndexError: list index out of range` there means that `radosgw-admin reshard list` returned an empty list

> 
> The [run](http://qa-proxy.ceph.com/teuthology/kapandya-2023-05-23_10:05:30-rgw-wip-rgw-yield-check-final-distro-default-smithi/7283499/teuthology.log) fails an s3-test in the crypt suite.

those kmip test failures are tracked in https://tracker.ceph.com/issues/58751

> 
> There are also some valgrind failures. Do any of these seem familiar @cbodley

i'm seeing two different valgrind reports on recent main, and opened https://tracker.ceph.com/issues/61428 and https://tracker.ceph.com/issues/61429 to track them. do those look the same?

**Comment by TRYTOBE8TME** [comment]
> > > Latest run: https://pulpito.ceph.com/kapandya-2023-05-23_10:05:30-rgw-wip-rgw-yield-check-final-distro-default-smithi/
> > 
> > 
> > That run looks cleaner but there are still a few failures I'm not sure are in the baseline: This [run](http://qa-proxy.ceph.com/teuthology/kapandya-2023-05-23_10:05:30-rgw-wip-rgw-yield-check-final-distro-default-smithi/7283525/teuthology.log) fails running rgw-reshard.sh in the workunits
> 
> that looks like a test case is failing: https://github.com/ceph/ceph/blob/62d1cc0/qa/workunits/rgw/test_rgw_reshard.py#L209-L212
> 
> the `IndexError: list index out of range` there means that `radosgw-admin reshard list` returned an empty list
> 

I don't see this failing in my latest run: https://pulpito.ceph.com/kapandya-2023-05-31_17:36:44-rgw-wip-rgw-yield-work-distro-default-smithi/

**Comment by cbodley** [comment]
> I don't see this failing in my latest run: https://pulpito.ceph.com/kapandya-2023-05-31_17:36:44-rgw-wip-rgw-yield-work-distro-default-smithi/

okay, strange. those results look good; the valgrind failures there are tracked in https://tracker.ceph.com/issues/61428

can you please rebase and rerun the suite to make sure the reshard failure is gone?

**Comment by mattbenjamin** [comment]
I see that make check failed:

[ RUN      ] Queue.SpawnAsyncRequest
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/rgw/test_rgw_dmclock_scheduler.cc:425: Failure
Value of: context.stopped()
  Actual: false
Expected: true
[  FAILED  ] Queue.SpawnAsyncRequest (9 ms)
[----------] 8 tests from Queue (42 ms total)

**Comment by TRYTOBE8TME** [comment]
Latest teuthology run: https://pulpito.ceph.com/soumyakoduri-2023-06-20_12:37:42-rgw-wip-rgw-yield-work-distro-default-smithi/

**Comment by TRYTOBE8TME** [comment]
jenkins test make check

**Comment by TRYTOBE8TME** [comment]
jenkins test api

**Comment by TRYTOBE8TME** [comment]
> I see that make check failed:
> 
> [ RUN ] Queue.SpawnAsyncRequest /home/jenkins-build/build/workspace/ceph-pull-requests/src/test/rgw/test_rgw_dmclock_scheduler.cc:425: Failure Value of: context.stopped() Actual: false Expected: true [ FAILED ] Queue.SpawnAsyncRequest (9 ms) [----------] 8 tests from Queue (42 ms total)

These went away but now I see ceph API tests check failing and it doesn't show any reason

**Comment by cbodley** [comment]
> Latest teuthology run: https://pulpito.ceph.com/soumyakoduri-2023-06-20_12:37:42-rgw-wip-rgw-yield-work-distro-default-smithi/

looks great! the s3test failure against dbstore is a new regression, and i've pushed a fix in https://github.com/ceph/s3-tests/pull/516. cc @soumyakoduri

**Comment by cbodley** [comment]
jenkins test api

**Comment by cbodley** [comment]
thanks @TRYTOBE8TME! are there more null_yields that we need to catch? @adamemerson had found some in https://github.com/ceph/ceph/pull/52036

as a next step, i'd like to find a way to add test coverage in teuthology to make sure that no requests are blocking. that might involve a config variable (something like `rgw_debug_require_beast_async`) that causes us to assert instead of just logging `WARNING: blocking librados call`. what do you think?

**Comment by yuvalif** [comment]
> thanks @TRYTOBE8TME! are there more null_yields that we need to catch? @adamemerson had found some in #52036
> 
> as a next step, i'd like to find a way to add test coverage in teuthology to make sure that no requests are blocking. that might involve a config variable (something like `rgw_debug_require_beast_async`) that causes us to assert instead of just logging `WARNING: blocking librados call`. what do you think?

maybe there should be a compilation flag to compile out all the blocking rados calls, and compile the RGW with this flag?

**Comment by TRYTOBE8TME** [comment]
> 

@cbodley yes there are some places where` null_yield` needs to be replaced and basically all those places are somewhere part of an op which is remaining to be done.

**Comment by cbodley** [comment]
> @cbodley does this PR need to block on adding tests which find _more_ blocking calls we could fix? I think we want to get this one in when it's complete and stable, no?

@mattbenjamin this PS has merged, sir! i think we're close to the finish line here, so i'd like to add some regression testing to catch any remaining blocking calls, and keep us from adding new ones. then we can do some performance measurements to figure out how to size the thread pool optimally

**Comment by cbodley** [comment]
> maybe there should be a compilation flag to compile out all the blocking rados calls, and compile the RGW with this flag?

@yuvalif we still need the blocking mode for radosgw-admin calls, librgw requests, and background threads. it's just the asio threads (like the beast frontend and `rgw::notify::Manager`) that should enforce async-only

**Comment by yuvalif** [comment]
> > maybe there should be a compilation flag to compile out all the blocking rados calls, and compile the RGW with this flag?
> 
> @yuvalif we still need the blocking mode for radosgw-admin calls, librgw requests, and background threads. it's just the asio threads (like the beast frontend and `rgw::notify::Manager`) that should enforce async-only

so, can we force an explicit `null_yield` in these cases?
or do you want tio catch beast+`null_yield` cases as well?

**Comment by cbodley** [comment]
i added this topic to next week's agenda https://pad.ceph.com/p/rgw-weekly

**Pull request merged**